Azure Arc ML training pipeline scenario with Blob

mdrakiburrahman opened the pull request:
Addresses [686](https://github.com/microsoft/azure_arc/issues/686).

likamrat commented on azure_arc_ml_jumpstart/aks/arm_template/aks.json:
```diff
@@ -0,0 +1,150 @@
+  "parameters": {
+    "clusterName": {
+      "type": "string",
+      "defaultValue": "Arc-ML-AKS",
```
Try to standardize on the "Demo" suffix and not the k8s distro.

mdrakiburrahman replied:
I'm not sure I'm following, this is directly in line with the [Data Services](https://github.com/microsoft/azure_arc/blob/53e89bbbfb96fbf2f231e1a321715ef81688f11d/azure_arc_data_jumpstart/aks/arm_template/aks.json#L7) scenarios, except 'Data' is replaced by 'ML`:
![image](https://user-images.githubusercontent.com/46581776/130327649-6c5057b6-b332-4fe1-8254-aa49ee8cd682.png)

likamrat replied:
I am aware, this is something we need to fix across all scenarios so it will be generic and we won't have to worry about infra type.

mdrakiburrahman replied:
Makes sense, changing to `Arc-ML`

likamrat commented on azure_arc_ml_jumpstart/aks/arm_template/artifacts/azuredeploy.parameters.example.json:
```diff
@@ -0,0 +1,39 @@
+    "templateBaseUrl": {
+      "value": "https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/"
+    },
+    "kubernetesVersion": {
+      "value": "1.19.11"
```
Were you able to test a higher version?

mdrakiburrahman replied:
Tested successfully on K8s 1.21.2:
![image](https://user-images.githubusercontent.com/46581776/130332757-ad2da5ce-db7b-4287-b162-9c312e8a36b6.png)

likamrat replied:
Thanks for verifying. Let's not change this for now as data services are still not supported with 1.2x and above

mdrakiburrahman replied:
Changing back to `1.19.11`

likamrat commented on azure_arc_ml_jumpstart/aks/arm_template/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,39 @@
+    "templateBaseUrl": {
+      "value": "https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/"
+    },
+    "kubernetesVersion": {
+      "value": "1.19.11"
```
see comment on version

mdrakiburrahman replied:
see above

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/_index.md:
```diff
@@ -0,0 +1,9 @@
+type: docs
+title: "Azure Arc-enabled machine learning"
+linkTitle: "Azure Arc-enabled machine learning"
+weight: 7
```
This should have _weight: 6_ and Lighthouse index should be 7

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
```diff
@@ -0,0 +1,238 @@
+
+![Deployed Architecture](./01.png)
+
+> **Note: Currently, Azure Arc-enabled Machine Learning is in [public preview](https://github.com/Azure/AML-Kubernetes)**.
```
```suggestion
> **Note: Currently, Azure Arc-enabled machine learning is in [public preview](https://github.com/Azure/AML-Kubernetes)**.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
```diff
@@ -0,0 +1,238 @@
+  * Model Training Run performed on Kubernetes Cluster - using [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/ml/job?view=azure-cli-latest#az_ml_job_create)
+  * [Model Registered](https://docs.microsoft.com/en-us/azure/machine-learning/concept-model-management-and-deployment#register-and-track-ml-models) on Azure ML Studio (via [`pickle`](https://docs.python.org/3/library/pickle.html) file)
+  * Model Inference endpoint (pod) deployed on Kubernetes Cluster - using [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/ml/endpoint?view=azure-cli-latest#az_ml_endpoint_create)
+  * Model Inference endpoint (pod) invoked using [1. Azure CLI](https://docs.microsoft.com/en-us/cli/azure/ml/endpoint?view=azure-cli-latest#az_ml_endpoint_invoke) and [2. PowerShell native RESTful methods](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-restmethod?view=powershell-7.1). The second method is to demonstrate that the deployed model can also be  ...
```
Do we want to incl. Postman in the Client VM?

mdrakiburrahman replied:
I don't think it'll add much value, since we're already showing how to call the RESTAPI via PowerShell. I've removed Postman verbiage in the latest commit to avoid any confusions.

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
````diff
@@ -0,0 +1,238 @@
+    For example:
+
+    ```shell
+    az group create --name Arc-Data-Demo --location "East US"
````
```suggestion
    az group create --name Arc-ML-Demo --location "East US"
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
````diff
@@ -0,0 +1,238 @@
+    --parameters templateBaseUrl="https://raw.githubusercontent.com/your--github--handle/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/"
+    ```
+
+    > **Note: The deployment time for this scenario can take ~8-10min**
````
Really?

mdrakiburrahman replied:
I was able to get an 8 minute deployment, the Bootstrap in ML is lighter than Data:
![8 min](https://raw.githubusercontent.com/microsoft/azure_arc/a34edef5b0e9c23ea135aa6775c9c53f695b49f5/docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/03.png)

That being said, changed readme to "10-15min" to avoid confusions

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
````diff
@@ -0,0 +1,238 @@
+
+    For example:
+
+    ```shell
+    az group create --name Arc-Data-Demo --location "East US"
+    az deployment group create \
+    --resource-group Arc-ML-Demo \
+    --name arcdata \
+    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/azuredeploy.json \
+    --parameters azuredeploy.parameters.json
+    --parameters templateBaseUrl="https://raw.githubusercontent.com/your--github--handle/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/"
+    ```
````
```suggestion
    ```shell
    az group create --name Arc-ML-Demo --location "East US"
    az deployment group create \
    --resource-group Arc-ML-Demo \
    --name arcml \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/azuredeploy.json \
    --parameters azuredeploy.parameters.json
    --parameters templateBaseUrl="https://raw.githubusercontent.com/your--github--handle/azure_arc/main/azure_arc_ml_jumpstart/aks/arm_template/"
    ```
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
````diff
@@ -0,0 +1,238 @@
+
+    > **Note: The deployment time for this scenario can take ~8-10min**
+
+* Once the above deployment is kicked off - you also have the option to deploy an Azure Bastion to access the Client VM. This is specially useful for enterprise environments where access to Port 3389 is restricted from Public IP's (e.g. using preventative [Azure Policies](https://docs.microsoft.com/en-us/azure/virtual-network/policy-reference#azure-virtual-network)).
+  
+  > **Note: Using Azure Bastion is completely optional for this scenario - if your environment allows RDP from your Public IP, feel free to skip this step**
+
+  To deploy Azure Bastion on the VNET (which should already be deployed via the ARM template above) - run the below commands:
+  ```shell
+    az network vnet subnet create -g Arc-ML-Demo --vnet-name "Arc-ML-VNet" -n "AzureBastionSubnet" --address-prefixes 172.16.2.0/27
+    az network public-ip create -g Arc-ML-Demo --name "Bastion-PIP" --sku Standard --location eastus
+    az network bastion create --name "Arc-Client-Bastion" --public-ip-address "Bastion-PIP" -g Arc-ML-Demo --vnet-name "Arc-ML-VNet" --location eastus
+  ```
````
nice addtion

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
```diff
@@ -0,0 +1,238 @@
+
+## Cluster extensions
+
+In this scenario, the Azure Arc-enabled Machine Learning services cluster extension was deployed and used throughout this scenario in order to deploy the machine learning components.
```
```suggestion
In this scenario, the Azure Arc-enabled machine learning services cluster extension was deployed and used throughout this scenario in order to deploy the machine learning components.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_ml/aks/aks_blob_mnist_arm_template/_index.md:
```diff
@@ -0,0 +1,238 @@
+  ![Delete Azure ML endpoints](./23.png)
+
+  ![Delete Azure ML endpoints](./24.png)
+  
```
```suggestion
```

likamrat reviewed (changes requested):
- Can we get rid of the dashes in the _simple-inference-cli.zip_ filename?
- Can we test 1.20.x AKS version?
- Can we get rid of the dashes in the _simple-train-cli.zip_ filename?
- Add _windowsConfiguration_ block in _clientVM.json_
- Update index weight
- Minor corrections

mdrakiburrahman commented:
@likamrat:
* Changed zip files with no dashes
* Tested on AKS v1.21.2
* Formatting suggestions
* Added logic for SP to work across multiple subscriptions

mdrakiburrahman commented:
- Changed AKS back to `1.19.11`
- Updated tag to `Arc-ML`

likamrat reviewed (approved):
LGTM. Great work!
